### ISSUE TITLE
Add code style analyzer to discourage null-forgiving operator (!) usage

### ISSUE THREAD

**Issue by santo998** [dotnet/roslyn#83946]
**Brief description:**

Add a configurable code style analyzer to discourage usage of the null-forgiving operator (`!`).

While the operator is useful in some scenarios, its usage appears to be increasingly overused in generated code, especially from Large Language Models (LLMs), where `!` is often added as a shortcut to silence nullable warnings instead of improving null-state analysis or introducing explicit null handling.

This can make nullable reference types less effective by hiding potential issues and normalizing warning suppression patterns. A configurable analyzer would help teams enforce stricter nullable practices and review these cases intentionally. The analyzer should be opt-in and configurable through `.editorconfig`.

**Languages applicable:**

C# only

**Code example that the analyzer should report:**

```csharp
string? nombre = GetName();

Console.WriteLine(name!.Length);
```

**Additional information:**

This should be a code style rule, not a correctness rule. Valid use cases for `!` exist, so the analyzer should be configurable and disabled by default, allowing teams to discourage or prohibit its usage according to their coding standards.

**Documentation requirements:**

When this analyzer is implemented, it must be documented by following the steps at Documentation for IDE CodeStyle analyzers.

**Comment by CyrusNajmabadi** [comment]
We already added an analyzer that tells you when you have an unneeded `!` and will remove it.

**Comment by santo998** [comment]
@CyrusNajmabadi Thanks, but I think this addresses a different problem than the one I am proposing.

The existing analyzer detects unnecessary usages of "!", i.e. cases where flow analysis already proves the value is non-null and the suppression is redundant.

My suggestion is about a configurable code style analyzer that can discourage or flag any usage of the null-forgiving operator, even when it is technically required to suppress a warning.

The motivation is not correctness, but maintainability and coding standards. In practice, "!" is increasingly being overused in generated code (especially from LLMs) as a quick way to silence nullable warnings instead of improving nullability annotations, adding explicit checks, or restructuring code.

For example, under the proposed rule this would still be reported intentionally:

Console.WriteLine(name!.Length);

even if "!" is semantically necessary to silence nullable analysis.

The goal would be similar to other configurable IDE style preferences: allow teams to say “we prefer avoiding "!" and want usages reviewed explicitly”, while still keeping legitimate cases possible through severity configuration or suppression.

**Comment by CyrusNajmabadi** [comment]
I don't think there is any chance of adding a warning for that. We don't add warnings for the normal and correct use of our constructs. You could write your own analyzer. I prompted Claude and it did it properly in one shot here.
